### PR TITLE
Fix Binomimal overflow when logits is large

### LINKED ISSUE #17843: Binomial.log_prob returns -inf when actual probability is 1 if logit is large
## 🐛 Bug

When a `Binomial` instance is initialized with a large logit, effectively setting the distribution mean to 1, `log_prob(1)` returns `-inf` whereas it should return `0`.

This is likely to occur when a neural net outputs the logit and tries to push for a probability of 1.

## To Reproduce

```
dist = torch.distributions.Binomial(logits=torch.Tensor([90.5229]))

dist.sample() 
>> tensor([1.])

dist.log_prob(dist.sample())
>> tensor([-inf])
```

## Expected behavior
Expected behavior should be as is its when initializing with smaller logits, such as:
```
dist = torch.distributions.Binomial(logits=torch.Tensor([15]))

dist.sample() 
>> tensor([1.])

dist.log_prob(dist.sample())
>> tensor([0.])
```

## Environment
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.15.1
[pip3] torch==0.4.1
[pip3] torchvision==0.2.1
[conda] Could not collect

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR fixes  #17843. In addition (test locally), this still maintains the continuity of log_prob which is addressed in https://github.com/pytorch/pytorch/pull/15962 

cc @neerajprad

### CODE REVIEW

**Comment by fehiepsi** [comment]
Although this is stable, the implementation will give incorrect grad when `logits=0` (only at this single point) due to `torch.abs` and `torch.clamp` operators. I'm not sure how to resolve it. We might reconsider the version pre-#15962.

**Comment by fehiepsi** [comment]
It turns out that the issue lies at torch.clamp which returns grad at 0 is 1. If it returns 0.5, grad will be correct.

Related https://github.com/pytorch/pytorch/pull/2195 where the same issue happens for binary cross entropy with logits. If we make grad of torch.clamp at 0 is 0.5, the result for grad at 0 of binomal logprob and binary cross entropy with logits (pre #2195) will be correct.  @apaszke What do you think?

**Comment by fehiepsi** [comment]
Nevermind, we have x = x.clamp(min=0) + x.clamp(max=0), the LHS has grad at 0 is 1 while RHS is 2. We can utilize this difference to add a dummy variable (x - x.clamp(min=0) - x.clamp(max=0)) / 2 to x.clamp(min=0) to adjust its grad.

**Comment by fehiepsi** [comment]
This is ready to review. I think that I have addressed problems which I can think about:
+ overflow of log_prob
+ discontinuity of grad of log_prob
+ correct grad at logits=0

**Review by neerajprad — APPROVED** [review]
Looks great!
